drivers/sensor/qdec_nxp_s32/qdec_nxp_s32.c

```diff
@@ -0,0 +1,424 @@
+/*
+ * Copyright 2023 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <stdint.h>
+
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/arch_interface.h>
+
+#ifndef M_PI
+#define M_PI 3.14159265358979323846
+#endif
+
+#define EMIOS_CHANNEL_COUNT 2U
+#define EMIOS_CW_CH_IDX     0
+#define EMIOS_CCW_CH_IDX    1
+
+/* LCU LUT control values for each of the 4 LC outputs */
+/* These values decide the direction of motor rotation */
+#define LCU_O0_LUT 0xAAAA
+#define LCU_O1_LUT 0xCCCC
+#define LCU_O2_LUT 0x4182
+#define LCU_O3_LUT 0x2814
+
+LOG_MODULE_REGISTER(nxp_qdec_s32, CONFIG_SENSOR_LOG_LEVEL);
+
+#include <Emios_Icu_Ip.h>
+#include <Trgmux_Ip.h>
+#include <Lcu_Ip.h>
+#include <Siul2_Port_Ip.h>
+
+#define DT_DRV_COMPAT nxp_qdec_s32
+
+#define EMIOS_NXP_S32_NODE(n) DT_NODELABEL(emios##n)
+
+/* Configuration variables from eMIOS Icu driver */
+extern eMios_Icu_Ip_ChStateType eMios_Icu_Ip_ChState[EMIOS_ICU_IP_NUM_OF_CHANNELS_USED];
+extern uint8 eMios_Icu_Ip_IndexInChState[EMIOS_ICU_IP_INSTANCE_COUNT][EMIOS_ICU_IP_NUM_OF_CHANNELS];
+
+static uint32_t Clockwise_overflow_count, CounterCW_overflow_count;
```

Comment:
Perhaps i missed something but look like those are not necessary, covered in here:

![image](https://github.com/zephyrproject-rtos/zephyr/assets/56692056/a4426dcf-83ba-4e55-9d7b-69f47cb7fd86)

Reply:
Removed from qdec_s32_data, as I am not sure how to pass any argument to Emios user notify callback, so using static variables for now.
But will need to add synchronization here, as these variables getting updated in interrupt callback and read in fetch.

Reply:
Added a semaphore

Reply:
Oh, now i see a semaphore can forever in interrupt context, which shouldn't (https://docs.zephyrproject.org/latest/kernel/services/synchronization/semaphores.html#c.k_sem_take)

![image](https://github.com/zephyrproject-rtos/zephyr/assets/56692056/fcc9b624-ad9c-4e4c-b248-4482a40ba4e6)

I was just asking, if is needed to avoid the race condition, you can justify to keep it

Reply:
@Dat-NguyenDuy  Thanks for pointing it out.. these callbacks are indeed in isr context, so changed the timeout to K_NO_WAIT

Reply:
Don't you need these for each instance? If you can't pass an argument to the callback APIs, then create instance-specific wrapper functions around the common callback function.